Set payment method and payment method title for Cash Orders

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12347
More context: pdfdoF-5ai-p2
<!-- Id number of the GitHub issue this PR addresses. -->

## Description

When collecting cash payments through the app, the order status is marked as `complete` without setting the `payment_method` and `payment_method_title` as its done on the web.

Cash payment logic is initiated by `PaymentMethodViewModel` through `markOrderAsPaidByCash` method. At the moment, all it does to the order is update its status to completed.

### Solution

1. Use `updateOrder` instead of `updateOrderStatus` action that allows us to set modified `Order` object, and pass an array of fields that should be updated on the backend.
2. Use `PaymentGateway.Constants.cashOnDeliveryGatewayID` (`cod`) as `paymentMethodID`
3. Fetch Cash on Delivery (COD) payment gateway through `viewStorage.loadPaymentGateway`. If it doesn't exist, set a default "Pay in Person" title

## Testing information

1. Create an order in the app
6. Open detail of the order
7. Tap on 'Collect Payment'
8. Select Cash
9. Confirm
10. Confirm the order is marked as completed, and payment_method and payment_method_title are set. Payment method title can be visible in order details or receipts.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/4062343/1a4e01df-c8ae-4a53-93ab-1580429a36e5

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Add support to update paymentMethodID when updating the order

staskus committed:
Set paymentMethodID when completing a cash payment

Use OrderAction.updateOrder instead of OrderAction.updateOrderStatus to update both status and paymentMethodID

Set paymentMethodID to "cod" when the cash payment is made

staskus committed:
Add support to update paymentMethodTitle when updating the order

staskus committed:
Set paymentMethodTitle when completing a cash payment

Use Cash On Delivery payment method title that is set on the payment method.

Since app supports cash payments regardless of Cash On Delivery settings on the backend, use a hardcoded title if the payment gateway or title doesn't exist

staskus committed:
Update OrdersRemote tests to confirm paymentMethodId and paymentMethodTitle are encoded

staskus committed:
Update PaymentMethodsViewModelTests to test setting paymentMethods for cash payments

staskus committed:
Remove unused code in OrderListViewModel

staskus committed:
Use cashOnDeliveryGatewayID instead of "cod" in the project

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Update RELEASE-NOTES.txt

staskus commented:
@iamgabrielma I still need to update UI test mocks to get the UI test to succeed. However, if you have time on Monday morning you can review the code and give some feedback, if not, we can do it once everything is completed, thanks! I'll also ask for further review from someone who worked on this part of the code.

staskus committed:
Update order with custom amount mock

When order is made with custom amount it is returned with auto-draft status, however when we want to collect payment, the status is updated to pending. We're using this order for custom amount payment so pending status is more appropriate for mocks and makes the order to be persisted in the db as expected.

staskus committed:
Increase priority of competed cash payment mock

Both complete_cash_simple_payment and orders_complete_custom_amount match the same endpoints. The default rule is that the matches are prioritized based on creation times. Until orders_complete_custom_amount was updated in the previous commit, complete_cash_simple_payment was prioritized higher.

Added explicit  highest priority to complete_cash_simple_payment so it would be prioritized higher for requests that match both.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented:
@toupper I've seen you have made the last larger improvements to cash payments (https://github.com/woocommerce/woocommerce-ios/pull/11365). Let me know if you would have time to take a look at the solution, at least from a high-level. Thanks!

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift:
```diff
@@ -338,13 +338,33 @@ private extension PaymentMethodsViewModel {
+                return
+            }
+
+            let cashOnDeliveryID = PaymentGateway.Constants.cashOnDeliveryGatewayID
```
I have my concerns hardcoding these title and payment gateway here. After all, this method is just marking the order as paid (complete); it doesn't have to be necessarily cash on delivery. I would change the function name to make it more explicit (e.g. `markOrderCompleteWithCashOnDelivery`), or pass the method id and title as parameters :)

staskus replied:
Good point. This method is only used for Cash Payments (Order -> Collect Payment -> Cash) flow, and is only called by a larger `markOrderAsPaidByCash` method. So I agree the naming is too broad and misleading. 👍

staskus replied:
Something like `completeCashPayment` or `completeCashOnDeliveryPayment` could reflect the functionality in a most appropriate way.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift:
```diff
@@ -338,13 +338,33 @@ private extension PaymentMethodsViewModel {
-                    return continuation.resume(returning: ())
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+            guard let order = ordersResultController.fetchedObjects.first else {
+                DDLogError("⛔️ Order not found, can't mark order as paid.")
```
Just for easier debugging, we could interpolate the orderID here:
```suggestion
                DDLogError("⛔️ Order \(orderID) not found, can't mark order as paid.")
```

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModelTests.swift:
```diff
@@ -135,18 +141,60 @@ final class PaymentMethodsViewModelTests: XCTestCase {
+    func test_mark_order_as_paid_by_cash_then_order_status_and_payment_method_fields_updated() async {
+        // Given
+        let stores = MockStoresManager(sessionManager: .testingInstance)
+        let storage = MockStorageManager()
```
Not necessarily for this PR, but since we have 20+ instances of `MockStorageManager()`, `siteID`, and others, we could make some `setUp` and `tearDown` in this test suite so each unit test is clearer 👍

staskus replied:
Good observation 👍 . Init and deinit of `MockStoresManager` and `MockStorageManage` can both be moved to setUp and tearDown. It's tougher with other parts of the tests since they differ slightly from one another.

iamgabrielma commented on Networking/NetworkingTests/Remote/OrdersRemoteTests.swift:
```diff
@@ -462,6 +462,21 @@ final class OrdersRemoteTests: XCTestCase {
         assertEqual(received, expected)
     }
 
+    func test_update_order_when_payment_method_id_and_title_passed_then_request_parameters_set() throws {
```
Thanks for covering the `OrdersRemoteTests` accounting for the new properties! 💯

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModelTests.swift:
```diff
@@ -135,18 +141,60 @@ final class PaymentMethodsViewModelTests: XCTestCase {
+        storage.insertSampleOrder(readOnlyOrder: order)
+        let paymentGateway = PaymentGateway.defaultPayInPersonGateway(siteID: siteID).copy(title: "Pay in Person")
+        storage.insertSamplePaymentGateway(readOnlyGateway: paymentGateway)
+        let dependencies = Dependencies(stores: stores, storage: storage)
```
I thought we wouldn't need to inject an instance of storage into the gateway dependencies for some cases, but that doesn't seem to be the case, if we do not specifically call `storage.insertSampleOrder(readOnlyOrder: .fake())` we'll see crashes in tests.

My understanding is this comes from the change from using `auto-draft` orders to `pending` orders, where auto-drafts are temporary and not meant to be stored, while pending orders are actual orders already started to be processed and should be stored. Does that sounds right?

If that's the case I think makes total sense for cash orders, but we should keep an eye to see if we've changed behaviour unintendedly where auto-drafts are used 👍

staskus replied:
We need to insert a sample order because we take an order from storage to pass to mody and then pass to `updateOrder` action. When we were using `updateOrderStatus`, we only needed orderID, so no order had to exist in the storage.

iamgabrielma reviewed (approved):
Works nicely! I've dropped a couple of comments and a question but nothing blocking, feel free to address them if you feel it's necessary or skip them 🙇 

LGTM! 🚢

staskus committed:
Merge branch 'trunk' into fix/12347-cash-orders-with-empty-payment-method

staskus committed:
Rename cash payment method to markOrderCompletedWithCashOnDelivery

The method is only used for Cash on Delivery payments

staskus committed:
Update logging when order cannot be found

staskus committed:
Move repeating init of services into setUp and tearDown

staskus committed:
Add separate mocks for auto_draft and pending cash payment state

- When custom amount is added to order, it's created in auto_draft state.
- When collect payment is tapped in the order, it's modified to pending state.
- When cash payment is done in the order, it's modified to complete state.
